### ISSUE TITLE
server.jetty.connection-idle-timeout unclear meaning

### ISSUE THREAD

**Issue by bernermic** [eclipse/jetty.project#6799]
**Jetty version**
jetty-9.4.43.v20210629

**Java version**
Java 16

**Question**
I have a Jetty (server) Web Application in which I want to stop requests after a timeot threshold is exceeded. 
According to Spring-Boot Docs for Jetty I configured the property [server.jetty.connection-idle-timeout](https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html#application-properties.server.server.jetty.connection-idle-timeout) and expected too long running request to be stopped with a server error.

Then I found [this reply to another issue](https://github.com/eclipse/jetty.project/issues/3261#issuecomment-454464197). I read this as one can not use this property in the jetty server, but it is only meant to be used in the jetty client. Do I get this right or am I missing something?

I made a failing test example here: https://github.com/bernermic/jetty-server-timeout-test which can be executed via `./mvnw verify`
The [SampleControllerIT](https://github.com/bernermic/jetty-server-timeout-test/blob/master/src/test/java/org/example/SampleControllerIT.java) sets the config property `server.jetty.connection-idle-timeout=1s` and expects some 5xx Status. Also I activated the loglevel `logging.level.org.eclipse.jetty.io=DEBUG`.
In the log I can see a `TimeoutException` and `Ignored idle endpoint ` message:
```
2021-09-13 17:36:00.809 DEBUG 759051 --- [uler-6b4283c4-1] org.eclipse.jetty.io.FillInterest        : onFail FillInterest@338c42fb{null}

java.util.concurrent.TimeoutException: Idle timeout expired: 1002/1000 ms
	at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:171) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:113) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:831) ~[na:na]

2021-09-13 17:36:00.809 DEBUG 759051 --- [uler-6b4283c4-1] org.eclipse.jetty.io.WriteFlusher        : ignored: WriteFlusher@2335a1a6{IDLE}->null

2021-09-13 17:36:00.810 DEBUG 759051 --- [uler-6b4283c4-1] org.eclipse.jetty.io.AbstractEndPoint    : Ignored idle endpoint SocketChannelEndPoint@167376e2{l=/127.0.0.1:38743,r=/127.0.0.1:37374,OPEN,fill=-,flush=-,to=4716/1000}{io=0/0,kio=0,kro=1}->HttpConnection@60e85226[p=HttpParser{s=CONTENT,0 of -1},g=HttpGenerator@3f4f541d{s=START}]=>HttpChannelOverHttp@9b9472b{s=HttpChannelState@11574b95{s=HANDLING rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=1,c=false/false,a=HANDLING,uri=//localhost:38743/,age=4713}
```

**Comment by sbordet** [comment]
In your example you `Thread.sleep()` inside the controller, which I assume it is called synchronously from `HttpServlet.service()`.

The Servlet specification requires that idle timeouts are ignored when threads are dispatched to `HttpServlet.service()`, unless they are doing blocking reads or blocking writes.

The behavior you see from Jetty is expected and correct.

**Comment by gregw** [comment]
To add some extra information....  the idle timeouts only apply to servlet container managed IO operation: waiting for a new request; reading a request; writing a response.    The servlet spec requires that if a thread is dispatched to the application, then it can take unlimited time  - even if that is a infinite loop or a sleep(forever).   The reason being that interrupting/stopping an application thread is not something that can be done  generally. Interrupts may be ignored or trigger bugs in the application code (frequently resulting in infinite loops) and stopping a thread is deprecated.

We can't even inform the application that the connection has closed underneath them as there is a limit of a single dispatched thread per request - ie we'd have to wait for the call to Servlet.service to return before we can call any onError callback.

I personally think the spec could have done more... but that is a horse that bolted decades ago.

I'm closing for now. Please reopen if you have further questions about this.
